arrow-data/src/data/types.rs

```diff
@@ -0,0 +1,153 @@
+pub enum DictionaryKeyType {
+    Int8,
+    Int16,
+    Int32,
+    Int64,
+    UInt8,
+    UInt16,
+    UInt32,
+    UInt64,
+}
+
+/// An enumeration of the types of run key
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub enum RunEndType {
+    Int16,
+    Int32,
+    Int64,
+}
+
+/// Describes the physical representation of a given [`DataType`]
+#[derive(Debug, Copy, Clone, Ord, PartialOrd, Eq, PartialEq, Hash)]
+pub enum PhysicalType {
+    Null,
+    Boolean,
+    Primitive(PrimitiveType),
+    FixedSizeBinary,
+    Bytes(OffsetType, BytesType),
+    FixedSizeList,
+    List(OffsetType),
+    Map,
+    Struct,
+    Union,
+    Dictionary(DictionaryKeyType),
+    Run(RunEndType),
+}
+
+impl From<&DataType> for PhysicalType {
+    fn from(value: &DataType) -> Self {
+        match value {
+            DataType::Null => Self::Null,
+            DataType::Boolean => Self::Boolean,
+            DataType::Int8 => Self::Primitive(PrimitiveType::Int8),
+            DataType::Int16 => Self::Primitive(PrimitiveType::Int16),
+            DataType::Int32 => Self::Primitive(PrimitiveType::Int32),
+            DataType::Int64 => Self::Primitive(PrimitiveType::Int64),
+            DataType::UInt8 => Self::Primitive(PrimitiveType::UInt8),
+            DataType::UInt16 => Self::Primitive(PrimitiveType::UInt16),
+            DataType::UInt32 => Self::Primitive(PrimitiveType::UInt32),
+            DataType::UInt64 => Self::Primitive(PrimitiveType::UInt64),
+            DataType::Float16 => Self::Primitive(PrimitiveType::Float16),
+            DataType::Float32 => Self::Primitive(PrimitiveType::Float32),
+            DataType::Float64 => Self::Primitive(PrimitiveType::Float64),
+            DataType::Timestamp(_, _) => Self::Primitive(PrimitiveType::Int64),
+            DataType::Date32 => Self::Primitive(PrimitiveType::Int32),
+            DataType::Date64 => Self::Primitive(PrimitiveType::Int64),
+            DataType::Time32(_) => Self::Primitive(PrimitiveType::Int32),
+            DataType::Time64(_) => Self::Primitive(PrimitiveType::Int64),
+            DataType::Duration(_) => Self::Primitive(PrimitiveType::Int64),
+            DataType::Decimal128(_, _) => Self::Primitive(PrimitiveType::Int128),
+            DataType::Decimal256(_, _) => Self::Primitive(PrimitiveType::Int256),
+            DataType::Interval(IntervalUnit::YearMonth) => {
+                Self::Primitive(PrimitiveType::Int32)
+            }
+            DataType::Interval(IntervalUnit::DayTime) => {
+                Self::Primitive(PrimitiveType::Int64)
+            }
+            DataType::Interval(IntervalUnit::MonthDayNano) => {
+                Self::Primitive(PrimitiveType::Int128)
+            }
+            DataType::Binary => Self::Bytes(OffsetType::Small, BytesType::Binary),
+            DataType::FixedSizeBinary(_) => Self::FixedSizeBinary,
+            DataType::LargeBinary => Self::Bytes(OffsetType::Large, BytesType::Binary),
```

Comment:
Just to keep variable length binary close.
```suggestion
            DataType::Binary => Self::Bytes(OffsetType::Small, BytesType::Binary),
            DataType::LargeBinary => Self::Bytes(OffsetType::Large, BytesType::Binary),
            DataType::FixedSizeBinary(_) => Self::FixedSizeBinary,
```